src/core/on_internal_error.cc

```diff
@@ -38,7 +38,7 @@ void seastar::on_internal_error(logger& logger, std::string_view msg) {
         logger.error("{}, at: {}", msg, current_backtrace());
         abort();
     } else {
-        logger.error(msg);
```

Comment:
I suspect that a lot of Seastar users will run into the same problem. I wonder if we can (and if we should?) fix it at the logger level, i.e., have a logger.error(string_view) overload which doesn't call fmt at all? This should make these changes unnecessary - here and in other Seastar-based projects

Reply:
@nyh hi Nadav, IMHO, to print a dynamic message as the only string is not very common use case. but if we really want to do this, yeah, as you suggest, we can add overloads for all variants of the logging helpers. like
```c++
#if FMT_VERSION >= 80000
    void error(std::string_view s) noexcept {
        log(log_level::error, fmt::runtime(s));
    }
#endif
``` 
the upside is that this allows Seastar and other seastar projects to compile with fmt v8 and up if they print runtime strings in this way. but the downside is that `log.hh` would be more segmented with `#if` macros.

Reply:
I think fmt::runtime(s) in your implementation is a mistake. It would mean that if the string s contains the characters "{}" for example, the logging will fail - while the intention of logger.error(s) was to print s as-is, without formatting. I think that error(std::string_view s) should print the string exactly. One easy thing you can do if we don't care about the performance of this supposedly-rare case:

```
#if FMT_VERSION >= 80000
    void error(std::string_view s) noexcept {
        log(log_level::error, "{}", s);
    }
#endif
```

I agree with you that it's not "a very common use case" but I'm guessing it is common enough that every Seastar project will contain a few of those and break on your original change. As you saw, Seastar itself had 5 of these cases. I just did a quick grep in the Scylla project and found there at least 5 such instances as well (there could be more, I wasn't very careful in my search).

If you can argue that for some reason doing logger.error(str) should be considered a mistake, and users should just do logger.error("{}", str) - then it's fine that we stop supporting this case, but if it's not really a mistake it would be unfortunate to have this non-backward-compatible change for no real reason.

I'm not sure I understand why you are bothered with these overloads staying forever. If we want to allow logger.error(s) to work forever, we need those overloads to stay forever, no?

Reply:
> I think `fmt::runtime(s)` in your implementation is a mistake. It would mean that if the string s contains the characters "{}" for example, the logging will fail - while the intention of logger.error(s) was to print s as-is, without formatting. I think that error(std::string_view s) should print the string exactly. One easy thing you can do if we don't care about the performance of this supposedly-rare case:
> 
> ```c++
> #if FMT_VERSION >= 80000
>     void error(std::string_view s) noexcept {
>         log(log_level::error, "{}", s);
>     }
> #endif
> ```

ahh, right! thought it was inconsistent that developers should:
```c++
logger.error("{{foo={},bar={}}}", foo, bar);
```
but when it comes to runtime string, they should use 
```c++
std::string m = "{foo=-1,bar=-1}";
logger.error(m);
```
but after staring at these code snippets for a while, i feel that the inconsistent usages are actually more intuitive.

> 
> I agree with you that it's not "a very common use case" but I'm guessing it is common enough that every Seastar project will contain a few of those and break on your original change. As you saw, Seastar itself had 5 of these cases. I just did a quick grep in the Scylla project and found there at least 5 such instances as well (there could be more, I wasn't very careful in my search).
> 
> If you can argue that for some reason doing logger.error(str) should be considered a mistake, and users should just do logger.error("{}", str) - then it's fine that we stop supporting this case, but if it's not really a mistake it would be unfortunate to have this non-backward-compatible change for no real reason.
> 

hmm, i am checking the good old `printf(3)` and friends. and even [python logging](https://docs.python.org/3/howto/logging.html#logging-basic-tutorial). i am convinced that it is not a mistake. we should not try to bend developers to a certain fmt function call.

> I'm not sure I understand why you are bothered with these overloads staying forever. If we want to allow logger.error(s) to work forever, we need those overloads to stay forever, no?

agreed. just didn't want to repeat this pattern for multiple times for each of the logging function. but it seems that's the right way to go.

all in all, i am convinced. will implement your suggestion.

@nyh thank you, Nadav!

Reply:
> ahh, right! thought it was inconsistent that developers should:
> 
> ```c++
> logger.error("{{foo={},bar={}}}", foo, bar);
> ```
> 
> but when it comes to runtime string, they should use
> 
> ```c++
> std::string m = "{foo=-1,bar=-1}";
> logger.error(m);
> ```

You saw less silly examples in Seastar itself which you needed to fix, and a less-silly example from ScyllaDB is:
```c++
logger.error(e.what());
```

I'm not saying that these examples are the epitome of good coding style. The code could - and maybe should - have done `logger.error("{}", e.what())`. But if we break this code that I guess most Seastar-based projects will need to fix at least a few places to fix this problem.

> hmm, i am checking the good old `printf(3)` and friends. and even [python logging](https://docs.python.org/3/howto/logging.html#logging-basic-tutorial). i am convinced that it is not a mistake. we should not try to bend developers to a certain fmt function call.

What is not a mistake - your proposal, or mine? I agree that your proposal is not a "mistake" in the sense that the old code probably behaved like in your proposal, i.e., before the patch, if you did
```
   std::string msg = "hello {} hi";
   logger.error(msg);
```
probably it used to generate a runtime error. Which is what your proposal also did, so in this sense, it's not a mistake. But I think that because there is no reasonable way to interpret this "{}" (because there are no parameters), I think we should just log the string as-is. I think this was just a bug waiting to happen - e.g., what if the logged string contains a file name and the file name has `{}` in it? So I think it's better to fix it now, **in addition** to keeping the old code working without change.

> all in all, i am convinced. will implement your suggestion.

Thanks!

Reply:
> What is not a mistake - your proposal, or mine?

sorry for the confusion. i was referring "If you can argue that for some reason doing logger.error(str) should be considered a mistake,". so  i meant to say `logger.error(str)` was not a mistake. =)